src/optimisation/Optimisation.jl

```diff
@@ -139,186 +154,364 @@ function (f::OptimLogDensity)(F, G, z)
+
+Generate an initial value for the optimization problem.
 
-function get_parameter_bounds(model::DynamicPPL.Model)
-    vi = DynamicPPL.VarInfo(model)
+If `initial_params` is not `nothing`, a copy of it is returned. Otherwise initial parameter
+values are generated either by sampling from the prior (if no constraints are present) or
+uniformly from the box constraints. If generic constraints are set, an error is thrown.
+"""
+function generate_initial_params(model::DynamicPPL.Model, initial_params, constraints)
+    if initial_params !== nothing
+        return copy(initial_params)
+    end
+    if has_generic_constraints(constraints)
+        throw(ArgumentError(
+            "You must provide an initial value when using generic constraints."
+        ))
+    end
+    if has_box_constraints(constraints)
+        return [
+            rand(Distributions.Uniform(lower, upper))
+            for (lower, upper) in zip(constraints.lb, constraints.ub)
+        ]
+    end
+    return rand(Vector, model)
+end
 
-    ## Check link status of vi
-    linked = DynamicPPL.istrans(vi)
-    
-    ## transform into unconstrained
-    if !linked
-        vi = DynamicPPL.link!!(vi, model)
+function default_solver(constraints::ModeEstimationConstraints)
+    return if has_generic_constraints(constraints)
+        OptimizationOptimJL.IPNewton()
+    else
+        OptimizationOptimJL.LBFGS()
     end
+end
 
-    d = length(vi[:])
-    lb = transform(fill(-Inf, d), vi, model, constrained_space{true}())
-    ub = transform(fill(Inf, d), vi, model, constrained_space{true}())
+"""
+    OptimizationProblem(log_density::OptimLogDensity, adtype, constraints)
 
-    return lb, ub
+Create an `OptimizationProblem` for the objective function defined by `log_density`.
+"""
+function Optimization.OptimizationProblem(log_density::OptimLogDensity, adtype, constraints)
+    # Note that OptimLogDensity is a callable that evaluates the model with given
+    # parameters. Hence we can use it in the objective function as below.
+    f = Optimization.OptimizationFunction(
+        (x, _log_density) -> _log_density(x), adtype; cons=constraints.cons
+    )
+    initial_params = log_density.varinfo[:]
+    prob = if !has_constraints(constraints)
+        Optimization.OptimizationProblem(f, initial_params, log_density)
+    else
+        Optimization.OptimizationProblem(
+            f, initial_params, log_density;
+            lcons=constraints.lcons,
+            ucons=constraints.ucons,
+            lb=constraints.lb,
+            ub=constraints.ub
+        )
+    end
+    return prob
 end
 
-function _optim_objective(model::DynamicPPL.Model, ::MAP, ::constrained_space{false})
-    ctx = OptimizationContext(DynamicPPL.DefaultContext())
-    obj = OptimLogDensity(model, ctx)
+"""
+    estimate_mode(
+        model::DynamicPPL.Model,
+        estimator::ModeEstimator,
+        [initial_params::Union{AbstractVector,Nothing}],
+        [solver];
+        <keyword arguments>
```

Comment:
Julia docs disagree though: https://docs.julialang.org/en/v1/manual/documentation/#Writing-Documentation

Reply:
Uhmm _interesting_ :thinking: 

Throughout TuringLang, we always use `; kwargs...` :grimacing: E.g. https://github.com/TuringLang/AbstractMCMC.jl/blob/2a77f53114c4b3c5a11a3759e39233dab9e16f9a/src/interface.jl#L63-L74

THough I was clearly very wrong here, I'd say do `kwargs...` instead of `<keyword arguments>` to be consistent with the rest of the code-bases in TuringLang. If we want to move to `<keyword arguments>`, then we should do that in a single push to ensure everything is consistent :+1: 

But thank you for making me aware of this; did not know this:)

Reply:
Makes sense, done.